alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/AbstractOrienting.kt

```diff
@@ -0,0 +1,417 @@
+                }, {
+                    /*
+                     * nearest door heuristic
+                     */
+                    crossingPoint(it).distanceTo(currentPosition)
+                })
+            )
+        if (minEdge != null) {
+            nextRoom = environment.graph().getEdgeTarget(minEdge)
+            subdestination = crossingPoint(minEdge)
+            targetDoor = minEdge
+            state = State.MOVING_TO_DOOR
+        }
+        /*
+         * Closed room, we can't move anywhere.
+         */
+        else {
+            state = State.ARRIVED
+        }
+    }
+
+    private fun moving() {
+        val inNewRoom = environment.graph().vertexSet()
+            .any {
+                (!::currRoom.isInitialized || it != currRoom) && it.contains(currentPosition)
+            }
+        if (inNewRoom) {
+            state = State.NEW_ROOM
+        } else if (state != State.CROSSING_DOOR) {
+            val arrived = currentPosition.distanceTo(subdestination) <= minDistance
+            if (state == State.MOVING_TO_DOOR) {
+                if (arrived) {
+                    /*
+                     * This may cause some unrealistic deviations, a possible alternative
+                     * is to move the pedestrian in the same direction it moved so far.
+                     */
+                    subdestination = nextRoom.centroid
+                    state = State.CROSSING_DOOR
+                } else {
+                    /*
+                     * Recomputes sub-destination
+                     */
+                    subdestination = crossingPoint(targetDoor)
+                }
+            } else if (arrived) {
+                state = State.ARRIVED
+            }
+        }
+    }
+
+    /*
+     * The orienting behavior is organised as a finite state machine, this method
+     * updates the internal state of this action but does not move the pedestrian.
+     * If the position of the pedestrian didn't change since the last call, this
+     * method has no effect.
+     */
+    private fun update() = when (state) {
+        State.START -> onStart()
+        State.NEW_ROOM -> inNewRoom()
+        State.MOVING_TO_DOOR, State.CROSSING_DOOR, State.MOVING_TO_FINAL -> moving()
+        State.ARRIVED -> {}
+    }
+
+    /**
+     * @returns the next position the pedestrians wants to reach.
+     */
+    override fun nextPosition(): P {
+        /*
+         * We perform an update so as to have the latest information possible, if the
+         * position of the pedestrian is always the same this will have no effect.
+         */
+        update()
+        return when (state) {
+            State.MOVING_TO_DOOR, State.CROSSING_DOOR, State.MOVING_TO_FINAL ->
+                interpolatePositions(currentPosition, subdestination)
+            /*
+             * The pedestrian doesn't want to move at all.
+             */
+            else -> currentPosition - currentPosition
```

Comment:
Ugh
if we don't have a "zero" or "origin" in `Vector`, then we must create it. We cannot subtract a vector from itself to get the origin, it's a kludge.